File: MonitorEdgeTunnelLib/Logger.cpp
#include "pch.h"
#include "Logger.h"
#include <iostream>

void Logger::SetLogCallback(LogCallback callback)
{
    GetInstance()._SetLogCallback(callback);
}

void Logger::Log(LogLevel level, std::string_view message)
{
    GetInstance()._Log(static_cast<int>(level), message);
}

void Logger::Log(int level, std::string_view message)
{
    GetInstance()._Log(level, message);
}

Logger& Logger::GetInstance()
{
    static Logger instance;
    return instance;
}

void Logger::_SetLogCallback(LogCallback callback)
{
    m_logCallback = callback;
}

void Logger::_Log(int level, std::string_view message)
{
    if (m_logCallback)
    {
        m_logCallback(level, message.data());
    }

    // Default logging behavior (e.g., output to console or file)
    std::cout << "Log [" << LogLevelEnumToString(level) << "]: " << message << std::endl;
}

std::string Logger::LogLevelEnumToString(int level) const
{
    switch (static_cast<LogLevel>(level))
    {
    case LogLevel::Debug:
        return "Debug";
    case LogLevel::Info:
        return "Info";
    case LogLevel::Warning:
        return "Warning";
    case LogLevel::Error:
        return "Error";
    case LogLevel::Fatal:
        return "Debug";
    default:
        return "Unknown";
    }
}

File: MonitorEdgeTunnelApp/Logger.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Serilog;

namespace MonitorEdgeTunnelApp
{
    public static class Logger
    {
        private static readonly string LogFilePath = System.IO.Path.Combine(System.AppDomain.CurrentDomain.BaseDirectory, "Log", "log.txt");

        // 建立 Serilog logger 實例
        private static readonly ILogger logger = new LoggerConfiguration()
            .WriteTo.Console()
            .WriteTo.File(LogFilePath, rollingInterval: RollingInterval.Day)
            .CreateLogger();

        public static void InitMonitorEdgeTunnelLogger()
        {
            MonitorEdgeTunnel.Instance.SetLogCallback((level, message) =>
            {
                Log((int)level, message);
            });
        }

        public static void DeinitMonitorEdgeTunnelLogger()
        {
            MonitorEdgeTunnel.Instance.SetLogCallback(null);
        }

        public static void Log(int level, string message)
        {
            // 根據 level 決定 log 級別
            switch (level)
            {
                case 0: // Debug
                    logger.Debug(message);
                    break;
                case 1: // Info
                    logger.Information(message);
                    break;
                case 2: // Warning
                    logger.Warning(message);
                    break;
                case 3: // Error
                    logger.Error(message);
                    break;
                case 4: // Fatal
                    logger.Fatal(message);
                    break;
                default:
                    break;
            }
        }
    }
}
